File: Code/Lesson_14_code/dns_query.py
"""A DNS query, built on the wire, with no library that hides the packet.

Lesson 1 read raw bytes off a TCP socket before naming the framework that
hides them. This is the same move for DNS: build the twelve-byte header and
the question by hand, send them over UDP to a resolver, and parse whatever
comes back byte by byte. `dig` and `socket.getaddrinfo` both do this; this
script shows what they do not print.

    python dns_query.py example.com A
    python dns_query.py example.com AAAA --server 1.1.1.1
    python dns_query.py github.com CNAME
    python dns_query.py google.com MX
    python dns_query.py example.com TXT
"""

from __future__ import annotations

import argparse
import random
import socket
import struct

TYPES = {"A": 1, "NS": 2, "CNAME": 5, "MX": 15, "TXT": 16, "AAAA": 28}
TYPE_NAMES = {v: k for k, v in TYPES.items()}


def encode_name(name: str) -> bytes:
    """A domain name on the wire is a sequence of length-prefixed labels,
    ending in a zero-length label. "example.com" becomes
    b"\\x07example\\x03com\\x00" — the length before each part is why a
    single label cannot exceed 63 bytes."""
    out = bytearray()
    for label in name.rstrip(".").split("."):
        out += bytes([len(label)]) + label.encode("ascii")
    out += b"\x00"
    return bytes(out)


def build_query(name: str, qtype: int) -> bytes:
    query_id = random.randint(0, 0xFFFF)
    # Header: id, flags (RD=1: "please recurse for me"), 1 question, 0 of
    # the rest.
    flags = 0x0100
    header = struct.pack(">HHHHHH", query_id, flags, 1, 0, 0, 0)
    question = encode_name(name) + struct.pack(">HH", qtype, 1)  # class IN
    return header + question


def read_name(packet: bytes, offset: int) -> tuple[str, int]:
    """A name later in the packet can point back into an earlier one
    instead of repeating it — a compression pointer, marked by the top two
    bits of a length byte both set. Follow it, but keep advancing the
    caller's cursor only past the pointer itself, not into the jump."""
    labels = []
    jumped = False
    end_offset = offset
    while True:
        length = packet[offset]
        if length & 0xC0 == 0xC0:
            pointer = struct.unpack(">H", packet[offset : offset + 2])[0] & 0x3FFF
            if not jumped:
                end_offset = offset + 2
                jumped = True
            offset = pointer
            continue
        if length == 0:
            offset += 1
            if not jumped:
                end_offset = offset
            break
        offset += 1
        labels.append(packet[offset : offset + length].decode("ascii"))
        offset += length
        if not jumped:
            end_offset = offset
    return ".".join(labels), end_offset


def parse_response(packet: bytes) -> None:
    query_id, flags, qdcount, ancount, nscount, arcount = struct.unpack(
        ">HHHHHH", packet[:12]
    )
    rcode = flags & 0x000F
    print(f"id={query_id:#06x} rcode={rcode} answers={ancount}")
    offset = 12
    for _ in range(qdcount):
        _, offset = read_name(packet, offset)
        offset += 4  # qtype + qclass
    for _ in range(ancount):
        name, offset = read_name(packet, offset)
        rtype, rclass, ttl, rdlength = struct.unpack(
            ">HHIH", packet[offset : offset + 10]
        )
        offset += 10
        rdata = packet[offset : offset + rdlength]
        type_name = TYPE_NAMES.get(rtype, str(rtype))
        if rtype == 1:  # A
            value = socket.inet_ntoa(rdata)
        elif rtype == 28:  # AAAA
            value = socket.inet_ntop(socket.AF_INET6, rdata)
        elif rtype in (5, 2):  # CNAME, NS
            value, _ = read_name(packet, offset)
        elif rtype == 15:  # MX
            preference = struct.unpack(">H", rdata[:2])[0]
            exchange, _ = read_name(packet, offset + 2)
            value = f"{preference} {exchange}"
        elif rtype == 16:  # TXT
            text_len = rdata[0]
            value = rdata[1 : 1 + text_len].decode("ascii", errors="replace")
        else:
            value = rdata.hex()
        print(f"{name}.\t{ttl}\tIN\t{type_name}\t{value}")
        offset += rdlength


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("name")
    parser.add_argument("type", choices=TYPES.keys())
    parser.add_argument("--server", default="1.1.1.1")
    parser.add_argument("--port", type=int, default=53)
    args = parser.parse_args()

    query = build_query(args.name, TYPES[args.type])
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.settimeout(5)
        sock.sendto(query, (args.server, args.port))
        response, _ = sock.recvfrom(4096)
    parse_response(response)


if __name__ == "__main__":
    main()

File: Code/js/Lesson_14_code/src/dnsQuery.ts
// A DNS query, built on the wire, with no library that hides the packet.
//
// The Python twin is dns_query.py — read it first; this is recognition,
// not new material. Same header, same length-prefixed labels, same
// compression pointers, over Node's `dgram` socket instead of Python's.
//
//   node dnsQuery.ts example.com A
//   node dnsQuery.ts example.com AAAA --server 1.1.1.1
//   node dnsQuery.ts www.github.com CNAME
//   node dnsQuery.ts google.com MX
//   node dnsQuery.ts example.com TXT

import dgram from "node:dgram";

const TYPES: Record<string, number> = { A: 1, NS: 2, CNAME: 5, MX: 15, TXT: 16, AAAA: 28 };
const TYPE_NAMES: Record<number, string> = Object.fromEntries(
  Object.entries(TYPES).map(([k, v]) => [v, k]),
);

function encodeName(name: string): Buffer {
  const parts: Buffer[] = [];
  for (const label of name.replace(/\.$/, "").split(".")) {
    parts.push(Buffer.from([label.length]), Buffer.from(label, "ascii"));
  }
  parts.push(Buffer.from([0]));
  return Buffer.concat(parts);
}

function buildQuery(name: string, qtype: number): Buffer {
  const id = Math.floor(Math.random() * 0xffff);
  const header = Buffer.alloc(12);
  header.writeUInt16BE(id, 0);
  header.writeUInt16BE(0x0100, 2); // RD = 1
  header.writeUInt16BE(1, 4); // 1 question
  const question = Buffer.concat([encodeName(name), Buffer.alloc(4)]);
  question.writeUInt16BE(qtype, question.length - 4);
  question.writeUInt16BE(1, question.length - 2); // class IN
  return Buffer.concat([header, question]);
}

function readName(packet: Buffer, offset: number): [string, number] {
  const labels: string[] = [];
  let jumped = false;
  let endOffset = offset;
  for (;;) {
    const length = packet[offset];
    if ((length & 0xc0) === 0xc0) {
      const pointer = packet.readUInt16BE(offset) & 0x3fff;
      if (!jumped) {
        endOffset = offset + 2;
        jumped = true;
      }
      offset = pointer;
      continue;
    }
    if (length === 0) {
      offset += 1;
      if (!jumped) endOffset = offset;
      break;
    }
    offset += 1;
    labels.push(packet.subarray(offset, offset + length).toString("ascii"));
    offset += length;
    if (!jumped) endOffset = offset;
  }
  return [labels.join("."), endOffset];
}

function parseResponse(packet: Buffer): void {
  const id = packet.readUInt16BE(0);
  const flags = packet.readUInt16BE(2);
  const qdcount = packet.readUInt16BE(4);
  const ancount = packet.readUInt16BE(6);
  const rcode = flags & 0x000f;
  console.log(`id=0x${id.toString(16).padStart(4, "0")} rcode=${rcode} answers=${ancount}`);

  let offset = 12;
  for (let i = 0; i < qdcount; i++) {
    [, offset] = readName(packet, offset);
    offset += 4;
  }
  for (let i = 0; i < ancount; i++) {
    const [name, afterName] = readName(packet, offset);
    offset = afterName;
    const rtype = packet.readUInt16BE(offset);
    const ttl = packet.readUInt32BE(offset + 4);
    const rdlength = packet.readUInt16BE(offset + 8);
    offset += 10;
    const rdata = packet.subarray(offset, offset + rdlength);
    const typeName = TYPE_NAMES[rtype] ?? String(rtype);
    let value: string;
    if (rtype === 1) value = Array.from(rdata).join(".");
    else if (rtype === 28) {
      const groups: string[] = [];
      for (let g = 0; g < 16; g += 2) groups.push(rdata.readUInt16BE(g).toString(16));
      value = groups.join(":");
    } else if (rtype === 5 || rtype === 2) [value] = readName(packet, offset);
    else if (rtype === 15) {
      const preference = rdata.readUInt16BE(0);
      const [exchange] = readName(packet, offset + 2);
      value = `${preference} ${exchange}`;
    } else if (rtype === 16) {
      const textLen = rdata[0];
      value = rdata.subarray(1, 1 + textLen).toString("ascii");
    } else value = rdata.toString("hex");
    console.log(`${name}.\t${ttl}\tIN\t${typeName}\t${value}`);
    offset += rdlength;
  }
}

async function main(): Promise<void> {
  const [name, type, ...rest] = process.argv.slice(2);
  const serverFlagIndex = rest.indexOf("--server");
  const server = serverFlagIndex >= 0 ? rest[serverFlagIndex + 1] : "1.1.1.1";
  const query = buildQuery(name, TYPES[type]);

  const socket = dgram.createSocket("udp4");
  const response = await new Promise<Buffer>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error("timed out")), 5000);
    socket.once("message", (msg) => {
      clearTimeout(timer);
      resolve(msg);
    });
    socket.send(query, 53, server);
  });
  socket.close();
  parseResponse(response);
}

main();
